### PR TITLE
Sep14 update

### PR DESCRIPTION
**Pull request by Stwissel**
## Changes

- added Eclipse (C) headers
- replaced generic JsonObject parameters with specific parameter objects
- moved db creation to Admin client
- split Create & update
- created helper methods for common Json operations
- simplify test runs

## Todo

- test coverage (test failing)
- replace StringFormat with UriTemplate, potentially target appendParamsToUrl for refactor
- documentation

### CODE REVIEW

File: src/main/java/io/vertx/ext/couchdb/CouchdbClient.java [hunk 1]
```diff
@@ -46,68 +61,215 @@ static CouchdbClient create(Vertx vertx, WebClient client, Credentials credentia
...
+   * @return Future with a CouchDbDatabase instance
    */
-  Future<JsonArray> dbsInfo(JsonObject options);
+  Future<CouchDbDatabase> getDatabase(String databaseName);
 
   /**
-   * Makes a call to the CouchDB server with the given parameters.
+   * @return the vertx instance
+   */
+  VertxInternal getVertx();
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:113]: Question Would `VertxInternal` be more beneficial for our users than returning `Vertx`?

File: src/main/java/io/vertx/ext/couchdb/admin/impl/CouchdbAdminImpl.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  @Override
+  public CouchdbClient getClient() {
+    return this.client;
+  }
+
+  @Override
+  public Future<JsonArray> activeTasks() {
+    String baseUrl = "/_active_tasks";
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/admin/impl/CouchdbAdminImpl.java:38]: Can the baseUrls be moved as constants?

Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/admin/impl/CouchdbAdminImpl.java:38]: It's a single place in the whole app. Not much gained

File: src/main/java/io/vertx/ext/couchdb/CouchdbClient.java (hunk 1 above)
Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:113]: As the name implies: for library internal use only

Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:113]: Removed in later PR

File: src/main/java/io/vertx/ext/couchdb/admin/impl/CouchdbAdminImpl.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    return this.client.getJsonArray(this.client.getWebClient(), baseUrl, null);
```
[the remark this answers is not in the document]
Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/admin/impl/CouchdbAdminImpl.java:39]: Is refactored in latest pr

File: src/main/java/io/vertx/ext/couchdb/CouchdbClient.java [hunk 4]
```diff
@@ -46,68 +61,215 @@ static CouchdbClient create(Vertx vertx, WebClient client, Credentials credentia
...
+
+  /**
+   * @return Credentials used
+   */
+  Credentials getCredentials();
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:118]: Should we still use the `get` prefix? I think we can use `credentials()` instead to be consistent with our other naming conventions. This also applies to other get methods here

File: src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java [hunk 5]
```diff
@@ -1,58 +1,185 @@
...
-    }
+    Objects.requireNonNull(docId);
+    Objects.requireNonNull(document);
+
+    Promise<JsonObject> promise = Promise.promise();
+    String urlToCheck = String.format("/%s/%s", this.databaseName, docId);
 
-    return client.rawCall(params).map(Buffer::toJsonObject);
+    this.client.doesExist(this.client.getWebClient(), urlToCheck)
+        .onSuccess(v -> promise.fail("Document alreday exists"))
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java:74]: Typo here on `already`

File: src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java [hunk 6]
```diff
@@ -1,58 +1,185 @@
...
-      .put("path", "/" + databaseName + "/" + docId)
-      .put("query", options);
+  public Future<JsonObject> updateDocument(String docId, String rev, JsonObject document) {
+
+    Objects.requireNonNull(docId);
+    Objects.requireNonNull(rev);
+    Objects.requireNonNull(document);
 
-    return client.rawCall(params).map(Buffer::toJsonObject);
+    if (document.containsKey("_rev") && !rev.equals(document.getString("_rev"))) {
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java:90]: Can we make `"_rev"` here a constant?

File: src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java [hunk 7]
```diff
@@ -1,58 +1,185 @@
...
 
   @Override
   public Future<JsonObject> status() {
-    throw new UnsupportedOperationException("Unimplemented method 'status'");
+    String urlToCheck = String.format("/%s/", databaseName);
+    return this.client.getJsonObject(this.client.getWebClient(), urlToCheck, null);
+  }
+
+  @Override
+  public Future<Buffer> getDocumentAttachment(String docId, String attachementName,
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java:128]: Fix typo here on `attachementName`

File: src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java [hunk 8]
```diff
@@ -1,58 +1,185 @@
...
+    }
+    UriTemplate template = UriTemplate.of("/{db}/{doc}/{attname}");
+    Variables variables = Variables.variables()
+        .set("db", this.databaseName)
+        .set("doc", docId)
+        .set("attname", attachementName);
+
+    String urlToCheck = params.appendParamsToUrl(template.expandToString(variables));
+
+    this.client.getWebClient().get(urlToCheck)
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java:145]: Should this code be on the client side like a `GET` request?

File: src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java [hunk 9]
```diff
@@ -1,58 +1,185 @@
...
+
+    Promise<JsonObject> promise = Promise.promise();
+
+    String urlToCheck = String.format("/%s/%s", this.databaseName, docId);
+    this.client.getEtag(this.client.getWebClient(), urlToCheck)
+        .onFailure(promise::fail)
+        .onSuccess(eTag -> {
+          if (force || rev.equals(eTag)) {
+
+            this.client.getWebClient().delete(urlToCheck)
```
Comment (jmgang) [inline src/main/java/io/vertx/ext/couchdb/database/impl/CouchDbDatabaseImpl.java:166]: Should this code be on the client like a DELETE request?

**Review by jmgang — CHANGES_REQUESTED** [review]
Hi @Stwissel , I got some questions and comments on this PR.

File: src/main/java/io/vertx/ext/couchdb/CouchdbClient.java (hunk 4 above)
Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/CouchdbClient.java:118]: Refactored away

File: src/main/java/io/vertx/ext/couchdb/utils/JsonArraySerializable.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.ext.couchdb.utils;
+
+import io.vertx.core.json.JsonArray;
+
+public interface JsonArraySerializable {
```
[the remark this answers is not in the document]
Comment (Stwissel) [reply src/main/java/io/vertx/ext/couchdb/utils/JsonArraySerializable.java:15]: Serializable is binary. This one says it has toJson for arrays

**Pull request merged**